homeassistant/components/stream/worker.py

```diff
@@ -0,0 +1,142 @@
+    audio_frame.sample_rate = 44100
+    audio_frame.time_base = Fraction(1, 44100)
+    return audio_frame
+
+
+def create_stream_buffer(stream_output, video_stream, audio_frame):
+    """Create a new StreamBuffer."""
+    import av
+    a_packet = None
+    segment = io.BytesIO()
+    output = av.open(
+        segment, mode='w', format=stream_output.format)
+    vstream = output.add_stream(
+        stream_output.video_codec, video_stream.rate)
+    # Fix format
+    vstream.codec_context.format = \
+        video_stream.codec_context.format
+    # Check if audio is requested
+    astream = None
+    if stream_output.audio_codec:
+        astream = output.add_stream(
+            stream_output.audio_codec, 44100)
+        # Need to do it multiple times for some reason
+        while not a_packet:
+            a_packets = astream.encode(audio_frame)
+            if a_packets:
+                a_packet = a_packets[0]
+    return (a_packet, StreamBuffer(segment, output, vstream, astream))
+
+
+def stream_worker(hass, stream, quit_event):
+    """Handle consuming streams."""
+    import av
+    try:
+        video_stream = stream.container.streams.video[0]
+    except (KeyError, IndexError):
+        hass.getLogger().error("Stream has no video")
+        return
+
+    audio_frame = generate_audio_frame()
+
+    outputs = {}
+    first_packet = True
+    sequence = 1
+    audio_packets = {}
+
+    while not quit_event.is_set():
+        try:
+            packet = next(stream.container.demux(video_stream))
+            if packet.dts is None:
+                # If we get a "flushing" packet, the stream is done
+                raise StopIteration
+        except (av.AVError, StopIteration):
+            # End of stream, clear listeners and stop thread
+            for fmt, buffer in outputs.items():
+                asyncio.run_coroutine_threadsafe(
+                    stream.outputs[fmt].put(None), hass.loop)
+            break
+
+        # Reset segment on every keyframe
+        if packet.is_keyframe:
+            # Save segment to outputs
+            segment_duration = (packet.pts * packet.time_base) / sequence
+            for fmt, buffer in outputs.items():
+                # Flush streams
+                for packet in buffer.vstream.encode():
+                    buffer.output.mux(packet)
+                if buffer.astream:
+                    for packet in buffer.astream.encode():
+                        buffer.output.mux(packet)
+                buffer.output.close()
+                del audio_packets[buffer.astream]
+                asyncio.run_coroutine_threadsafe(
+                    stream.outputs[fmt].put(Segment(
+                        sequence, buffer.segment, segment_duration
+                    )), hass.loop)
+
+            # Clear outputs and increment sequence
+            outputs = {}
+            if not first_packet:
+                sequence += 1
+
+            # Initialize outputs
+            if not outputs:
```

Comment:
You just cleared `outputs` several lines above, why check outputs here?
If my understanding correct, you are creating new buffer for each key frame, what is the purpose of this?

Reply:
The way HLS works, a single stream is segmented into multiple files.  The worker performs this logic and saves it to an `io.BytesIO()` object.  That makes it easy to render each requested segment stream for HLS, and the video segments can be combined for other formats like progressive mp4 download or MJPEG stream in the future.  This also allows us to discard old segments and save on system resources.

As far as the sanity check on `outputs`, I believe that is carry over from when I did not have the two streams of logic in the same conditional block, and can probably be safely removed.